From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: schr-0dinger <kmkailasnath@gmail.com>
Date: Sun, 16 Aug 2026 07:38:36 +0530
Subject: [PATCH 03/15] imsservice: Update to a later patched revision

From schr-0dinger's "update IMS blobs from working device" (vendor
bb3ffd6), a later revision of the patched imsservice.apk:

- GoogleImsService.getPendingCallSession(): skip a pending session of
  the other SIM slot.
- GoogleImsService$ImsNotifier: drop the onCdpnInfo() notification.
- imsprofile.json: SIP over TCP for RJIL VoLTE and RJIL Emergency.

Assisted-by: LLM
---
 res/raw/imsprofile.json                       |   4 +-
 .../ims/GoogleImsService$ImsNotifier.smali    |  65 +++-------
 smali/com/google/ims/GoogleImsService.smali   | 112 +++++++++++++-----
 3 files changed, 98 insertions(+), 83 deletions(-)

diff --git a/res/raw/imsprofile.json b/res/raw/imsprofile.json
index 4cf8edc..a751754 100644
--- a/res/raw/imsprofile.json
+++ b/res/raw/imsprofile.json
@@ -16064,7 +16064,7 @@
       "need_ipv4_dns": true,
       "auth_algo": "hmac-md5-96",
       "pdn": "ims",
-      "transport": "udp",
+      "transport": "tcp",
       "mss_size": 1300,
       "priority": 1,
       "support_ipsec": true,
@@ -16153,7 +16153,7 @@
       "auth_algo": "hmac-md5-96",
       "pdn": "emergency",
       "emergency_support": true,
-      "transport": "udp",
+      "transport": "tcp",
       "mss_size": 1300,
       "priority": 1,
       "support_ipsec": true,
diff --git a/smali/com/google/ims/GoogleImsService$ImsNotifier.smali b/smali/com/google/ims/GoogleImsService$ImsNotifier.smali
index a0661a6..bd38bfd 100644
--- a/smali/com/google/ims/GoogleImsService$ImsNotifier.smali
+++ b/smali/com/google/ims/GoogleImsService$ImsNotifier.smali
@@ -955,41 +955,11 @@
     invoke-static {v1, v2}, Landroid/util/Log;->d(Ljava/lang/String;Ljava/lang/String;)I
 
     .line 1526
-    :try_start_0
-    iget-object v1, p0, Lcom/google/ims/GoogleImsService$ImsNotifier;->this$0:Lcom/google/ims/GoogleImsService;
-
-    invoke-static {v1}, Lcom/google/ims/GoogleImsService;->access$500(Lcom/google/ims/GoogleImsService;)Ljava/util/Map;
-
-    move-result-object v1
-
-    invoke-static {p1}, Ljava/lang/Integer;->valueOf(I)Ljava/lang/Integer;
-
-    move-result-object v2
-
-    invoke-interface {v1, v2}, Ljava/util/Map;->get(Ljava/lang/Object;)Ljava/lang/Object;
-
-    move-result-object v1
-
-    check-cast v1, Lcom/android/ims/internal/ISecImsMmTelEventListener;
-
-    invoke-interface {v1, p2, p3}, Lcom/android/ims/internal/ISecImsMmTelEventListener;->onCdpnInfo(Ljava/lang/String;I)V
-    :try_end_0
-    .catch Landroid/os/RemoteException; {:try_start_0 .. :try_end_0} :catch_0
-
-    .line 1529
-    goto :goto_0
+    const-string/jumbo v2, "onCdpnInfo: skip ISecImsMmTelEventListener on AOSP"
 
-    .line 1527
-    :catch_0
-    move-exception v1
-
-    .line 1528
-    .local v1, "e":Landroid/os/RemoteException;
-    invoke-virtual {v1}, Landroid/os/RemoteException;->printStackTrace()V
+    invoke-static {v1, v2}, Landroid/util/Log;->d(Ljava/lang/String;Ljava/lang/String;)I
 
     .line 1530
-    .end local v1    # "e":Landroid/os/RemoteException;
-    :goto_0
     return-void
 .end method
 
@@ -1282,6 +1252,10 @@
 
     invoke-virtual {v5, v6, v3}, Landroid/content/Intent;->putExtra(Ljava/lang/String;Ljava/io/Serializable;)Landroid/content/Intent;
 
+    const-string v6, "android:phone_id"
+
+    invoke-virtual {v5, v6, p1}, Landroid/content/Intent;->putExtra(Ljava/lang/String;I)Landroid/content/Intent;
+
     .line 1473
     :try_start_0
     iget-object v6, p0, Lcom/google/ims/GoogleImsService$ImsNotifier;->this$0:Lcom/google/ims/GoogleImsService;
@@ -1530,29 +1504,25 @@
     .end local v6    # "e":Landroid/os/RemoteException;
     :goto_2
     :try_start_1
-    iget-object v6, p0, Lcom/google/ims/GoogleImsService$ImsNotifier;->this$0:Lcom/google/ims/GoogleImsService;
-
-    invoke-static {v6}, Lcom/google/ims/GoogleImsService;->access$500(Lcom/google/ims/GoogleImsService;)Ljava/util/Map;
+    invoke-virtual {v4}, Lcom/google/ims/internal/ServiceProfile;->getIncomingCallIntent()Landroid/app/PendingIntent;
 
     move-result-object v6
 
-    invoke-static {p1}, Ljava/lang/Integer;->valueOf(I)Ljava/lang/Integer;
-
-    move-result-object v7
+    if-eqz v6, :cond_6
 
-    invoke-interface {v6, v7}, Ljava/util/Map;->get(Ljava/lang/Object;)Ljava/lang/Object;
+    const-wide/16 v7, 0x1f4
 
-    move-result-object v6
+    invoke-static {v7, v8}, Landroid/os/SystemClock;->sleep(J)V
 
-    check-cast v6, Lcom/android/ims/internal/ISecImsMmTelEventListener;
+    iget-object v7, p0, Lcom/google/ims/GoogleImsService$ImsNotifier;->this$0:Lcom/google/ims/GoogleImsService;
 
-    invoke-virtual {v5}, Landroid/content/Intent;->getExtras()Landroid/os/Bundle;
+    invoke-virtual {v7}, Lcom/google/ims/GoogleImsService;->getContext()Landroid/content/Context;
 
     move-result-object v7
 
-    invoke-interface {v6, p2, v7}, Lcom/android/ims/internal/ISecImsMmTelEventListener;->onIncomingCall(ILandroid/os/Bundle;)V
+    invoke-virtual {v6, v7, v0, v5}, Landroid/app/PendingIntent;->send(Landroid/content/Context;ILandroid/content/Intent;)V
     :try_end_1
-    .catch Landroid/os/RemoteException; {:try_start_1 .. :try_end_1} :catch_1
+    .catch Landroid/app/PendingIntent$CanceledException; {:try_start_1 .. :try_end_1} :catch_1
 
     .line 1513
     goto :goto_3
@@ -1563,10 +1533,7 @@
 
     .line 1511
     .restart local v6    # "e":Landroid/os/RemoteException;
-    invoke-virtual {v6}, Landroid/os/RemoteException;->printStackTrace()V
-
-    .line 1512
-    invoke-static {v0}, Ljava/lang/System;->exit(I)V
+    invoke-virtual {v6}, Landroid/app/PendingIntent$CanceledException;->printStackTrace()V
 
     .line 1515
     .end local v3    # "key":Ljava/lang/Integer;
@@ -1574,11 +1541,11 @@
     .end local v5    # "fillIn":Landroid/content/Intent;
     .end local v6    # "e":Landroid/os/RemoteException;
     :cond_6
-    :goto_3
     goto/16 :goto_0
 
     .line 1516
     :cond_7
+    :goto_3
     return-void
 .end method
 
diff --git a/smali/com/google/ims/GoogleImsService.smali b/smali/com/google/ims/GoogleImsService.smali
index 8dee020..7a7c7e1 100644
--- a/smali/com/google/ims/GoogleImsService.smali
+++ b/smali/com/google/ims/GoogleImsService.smali
@@ -3576,11 +3576,11 @@
 
     const-string v4, "GoogleImsService"
 
-    if-eqz v3, :cond_e
+    if-eqz v3, :cond_10
 
     iget-object v3, v0, Lcom/google/ims/GoogleImsService;->mVolteServiceModule:Lcom/sec/internal/ims/volte2/VolteServiceModule;
 
-    if-eqz v3, :cond_e
+    if-eqz v3, :cond_10
 
     .line 540
     move-object/from16 v5, p2
@@ -3591,6 +3591,54 @@
 
     .line 541
     .local v3, "session":Lcom/sec/ims/volte2/IImsCallSession;
+    if-eqz v3, :cond_0
+
+    invoke-interface {v3}, Lcom/sec/ims/volte2/IImsCallSession;->getPhoneId()I
+
+    move-result v5
+
+    if-eq v5, v2, :cond_1
+
+    new-instance v6, Ljava/lang/StringBuilder;
+
+    invoke-direct {v6}, Ljava/lang/StringBuilder;-><init>()V
+
+    const-string/jumbo v7, "getPendingCallSession: service phoneId "
+
+    invoke-virtual {v6, v7}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    invoke-virtual {v6, v2}, Ljava/lang/StringBuilder;->append(I)Ljava/lang/StringBuilder;
+
+    const-string/jumbo v7, " != pending session phoneId "
+
+    invoke-virtual {v6, v7}, Ljava/lang/StringBuilder;->append(Ljava/lang/String;)Ljava/lang/StringBuilder;
+
+    invoke-virtual {v6, v5}, Ljava/lang/StringBuilder;->append(I)Ljava/lang/StringBuilder;
+
+    invoke-virtual {v6}, Ljava/lang/StringBuilder;->toString()Ljava/lang/String;
+
+    move-result-object v6
+
+    invoke-static {v4, v6}, Landroid/util/Log;->e(Ljava/lang/String;Ljava/lang/String;)I
+
+    new-instance v7, Landroid/os/RemoteException;
+
+    invoke-direct {v7, v6}, Landroid/os/RemoteException;-><init>(Ljava/lang/String;)V
+
+    throw v7
+
+    :cond_0
+    const-string/jumbo v5, "getPendingCallSession: no pending session"
+
+    invoke-static {v4, v5}, Landroid/util/Log;->e(Ljava/lang/String;Ljava/lang/String;)I
+
+    new-instance v6, Landroid/os/RemoteException;
+
+    invoke-direct {v6, v5}, Landroid/os/RemoteException;-><init>(Ljava/lang/String;)V
+
+    throw v6
+
+    :cond_1
     invoke-static {}, Lcom/sec/internal/ims/volte2/ImsCallSession$STATE;->values()[Lcom/sec/internal/ims/volte2/ImsCallSession$STATE;
 
     move-result-object v6
@@ -3605,11 +3653,11 @@
     .local v6, "sessionState":Lcom/sec/internal/ims/volte2/ImsCallSession$STATE;
     sget-object v7, Lcom/sec/internal/ims/volte2/ImsCallSession$STATE;->EndingCall:Lcom/sec/internal/ims/volte2/ImsCallSession$STATE;
 
-    if-eq v6, v7, :cond_d
+    if-eq v6, v7, :cond_f
 
     sget-object v7, Lcom/sec/internal/ims/volte2/ImsCallSession$STATE;->EndedCall:Lcom/sec/internal/ims/volte2/ImsCallSession$STATE;
 
-    if-eq v6, v7, :cond_d
+    if-eq v6, v7, :cond_f
 
     .line 547
     invoke-interface {v3}, Lcom/sec/ims/volte2/IImsCallSession;->getCallProfile()Lcom/sec/ims/volte2/data/CallProfile;
@@ -3654,7 +3702,7 @@
 
     .line 553
     .local v9, "registration":Lcom/sec/ims/ImsRegistration;
-    if-eqz v9, :cond_3
+    if-eqz v9, :cond_5
 
     .line 554
     invoke-virtual {v9}, Lcom/sec/ims/ImsRegistration;->getCurrentRat()I
@@ -3667,7 +3715,7 @@
 
     move-result v13
 
-    if-eqz v13, :cond_0
+    if-eqz v13, :cond_2
 
     .line 556
     invoke-virtual {v7}, Lcom/sec/ims/volte2/data/CallProfile;->getRadioTech()I
@@ -3675,7 +3723,7 @@
     move-result v11
 
     .line 558
-    :cond_0
+    :cond_2
     invoke-interface {v3}, Lcom/sec/ims/volte2/IImsCallSession;->getMdecType()I
 
     move-result v13
@@ -3684,7 +3732,7 @@
 
     const-string v15, "CallRadioTech"
 
-    if-ne v13, v14, :cond_1
+    if-ne v13, v14, :cond_3
 
     .line 559
     const-string/jumbo v13, "set EXTRA_CALL_RAT_TYPE as RIL_RADIO_TECHNOLOGY_LTE due to call icon in callend screen."
@@ -3703,7 +3751,7 @@
     goto :goto_0
 
     .line 562
-    :cond_1
+    :cond_3
     invoke-static {v11}, Ljava/lang/String;->valueOf(I)Ljava/lang/String;
 
     move-result-object v13
@@ -3714,13 +3762,13 @@
     :goto_0
     const/16 v13, 0x12
 
-    if-ne v11, v13, :cond_2
+    if-ne v11, v13, :cond_4
 
     move v13, v12
 
     goto :goto_1
 
-    :cond_2
+    :cond_4
     const/4 v13, 0x0
 
     :goto_1
@@ -3728,7 +3776,7 @@
 
     .line 567
     .end local v11    # "currentRat":I
-    :cond_3
+    :cond_5
     invoke-virtual {v7}, Lcom/sec/ims/volte2/data/CallProfile;->getDialingNumber()Ljava/lang/String;
 
     move-result-object v11
@@ -3765,13 +3813,13 @@
 
     move-result v14
 
-    if-eqz v14, :cond_4
+    if-eqz v14, :cond_6
 
     .line 574
     const-string/jumbo v11, "unknown"
 
     .line 577
-    :cond_4
+    :cond_6
     invoke-virtual {v9}, Lcom/sec/ims/ImsRegistration;->getImsProfile()Lcom/sec/ims/settings/ImsProfile;
 
     move-result-object v14
@@ -3792,7 +3840,7 @@
     .local v15, "oir":I
     sget-object v10, Lcom/sec/internal/ims/util/Mno;->DOCOMO:Lcom/sec/internal/ims/util/Mno;
 
-    if-ne v14, v10, :cond_5
+    if-ne v14, v10, :cond_7
 
     .line 580
     invoke-virtual {v0, v13}, Lcom/google/ims/GoogleImsService;->getOirExtraFromDialingNumberForDcm(Ljava/lang/String;)I
@@ -3806,7 +3854,7 @@
     .line 582
     .end local v10    # "oir":I
     .restart local v15    # "oir":I
-    :cond_5
+    :cond_7
     invoke-virtual {v0, v11}, Lcom/google/ims/GoogleImsService;->getOirExtraFromDialingNumber(Ljava/lang/String;)I
 
     move-result v10
@@ -3923,7 +3971,7 @@
 
     move-result-object v1
 
-    if-eqz v1, :cond_7
+    if-eqz v1, :cond_9
 
     .line 597
     invoke-virtual {v7}, Lcom/sec/ims/volte2/data/CallProfile;->getHistoryInfo()Ljava/lang/String;
@@ -3947,7 +3995,7 @@
 
     const-string v12, "com.samsung.telephony.extra.CALL_FORWARDING_PRESENTATION"
 
-    if-eqz v1, :cond_6
+    if-eqz v1, :cond_8
 
     .line 599
     invoke-virtual {v8, v12, v15}, Landroid/telephony/ims/ImsCallProfile;->setCallExtra(Ljava/lang/String;Ljava/lang/String;)V
@@ -3955,19 +4003,19 @@
     goto :goto_3
 
     .line 601
-    :cond_6
+    :cond_8
     const-string v1, "0"
 
     invoke-virtual {v8, v12, v1}, Landroid/telephony/ims/ImsCallProfile;->setCallExtra(Ljava/lang/String;Ljava/lang/String;)V
 
     .line 605
-    :cond_7
+    :cond_9
     :goto_3
     invoke-virtual {v7}, Lcom/sec/ims/volte2/data/CallProfile;->getVerstat()Ljava/lang/String;
 
     move-result-object v1
 
-    if-eqz v1, :cond_8
+    if-eqz v1, :cond_a
 
     .line 608
     invoke-virtual {v7}, Lcom/sec/ims/volte2/data/CallProfile;->getVerstat()Ljava/lang/String;
@@ -3979,14 +4027,14 @@
     invoke-virtual {v8, v12, v1}, Landroid/telephony/ims/ImsCallProfile;->setCallExtra(Ljava/lang/String;Ljava/lang/String;)V
 
     .line 611
-    :cond_8
+    :cond_a
     invoke-virtual {v7}, Lcom/sec/ims/volte2/data/CallProfile;->getHDIcon()I
 
     move-result v1
 
     const/4 v12, 0x1
 
-    if-ne v1, v12, :cond_9
+    if-ne v1, v12, :cond_b
 
     .line 612
     const/4 v1, 0x0
@@ -3996,7 +4044,7 @@
     goto :goto_4
 
     .line 614
-    :cond_9
+    :cond_b
     const/4 v1, 0x0
 
     const/4 v12, 0x3
@@ -4011,7 +4059,7 @@
 
     move-result v12
 
-    if-eqz v12, :cond_c
+    if-eqz v12, :cond_e
 
     .line 618
     new-instance v12, Landroid/os/Bundle;
@@ -4082,7 +4130,7 @@
     .line 625
     const/4 v2, 0x1
 
-    if-ne v15, v2, :cond_b
+    if-ne v15, v2, :cond_d
 
     .line 626
     invoke-static {}, Lcom/sec/internal/ims/imsservice/ImsServiceStub;->getInstance()Lcom/sec/internal/ims/imsservice/ImsServiceStub;
@@ -4101,13 +4149,13 @@
     .local v2, "mdecLineSlotIndex":I
     const/4 v5, -0x1
 
-    if-ne v2, v5, :cond_a
+    if-ne v2, v5, :cond_c
 
     const/16 v16, 0x0
 
     goto :goto_5
 
-    :cond_a
+    :cond_c
     move/from16 v16, v2
 
     :goto_5
@@ -4145,7 +4193,7 @@
     .end local v2    # "mdecLineSlotIndex":I
     .end local v16    # "sessionId":I
     .restart local v1    # "sessionId":I
-    :cond_b
+    :cond_d
     move/from16 v16, v1
 
     .line 631
@@ -4166,7 +4214,7 @@
     .end local v16    # "sessionId":I
     .end local v18    # "phoneId":I
     .local v2, "phoneId":I
-    :cond_c
+    :cond_e
     move/from16 v18, v2
 
     .line 634
@@ -4208,7 +4256,7 @@
     .end local v18    # "phoneId":I
     .local v1, "service":Lcom/google/ims/internal/ServiceProfile;
     .restart local v2    # "phoneId":I
-    :cond_d
+    :cond_f
     move-object/from16 v17, v1
 
     move/from16 v18, v2
@@ -4236,7 +4284,7 @@
     .end local v18    # "phoneId":I
     .restart local v1    # "service":Lcom/google/ims/internal/ServiceProfile;
     .restart local v2    # "phoneId":I
-    :cond_e
+    :cond_10
     move-object/from16 v17, v1
 
     move/from16 v18, v2
